### PR TITLE
[Clang][Sema] Diagnose unexpanded packs in the template argument lists of function template specializations

### PR DESCRIPTION
**Pull request by sdkrystian**
This diagnoses unexpanded packs in the _unqualified-id_ of a function template specialization's _declarator-id_. For example:
```cpp
template<typename... Ts>
struct A
{
    template<typename U>
    void f();

    template<>
    void f<Ts>(); // error: explicit specialization contains unexpanded parameter pack 'Ts'
};
```

I moved the handling of template-id's so it happens right after we determine whether we are declaring a function template/function template specialization so diagnostics are issued in lexical order.

### CODE REVIEW

**Comment by sdkrystian** [comment]
Ping @erichkeane

**Comment by cor3ntin** [comment]
Can you add a release note?
Otherwise, I think the changes are reasonable. I'd like a second pair of eyes though @erichkeane

**Review by erichkeane — COMMENTED** [review]
As Corentin said, needs a release note, else LGTM.

File: clang/docs/ReleaseNotes.rst
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:521]: Do you have a github issue that this fixes, or is this just something that never had an issue?  If it has one, please include here, else, this is fine.

**Review by erichkeane — APPROVED** [review]

File: clang/docs/ReleaseNotes.rst
Comment (sdkrystian) [reply clang/docs/ReleaseNotes.rst:521]: @erichkeane I just looked -- I don't see an open issue.

**Comment by sdkrystian** [comment]
Added release note & rebased. I'll need someone to merge this for me :)

**Comment by nico** [comment]
This broke check-clang-tools: http://45.33.8.238/linux/127018/step_8.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by erichkeane** [comment]
Reverted here: https://github.com/llvm/llvm-project/pull/76876

@sdkrystian : please take a look at the regression that nico pointed out and re-submit with considerations for that.

**Pull request merged**
